Feature: Numeric separators with underscores for improved readability

## Feature Description

Add support for underscores as visual separators in numeric literals to improve readability of large numbers, following conventions from Rust, Python, Java, and other modern languages.

## Motivation

Large numeric literals can be difficult to read:
```ez
temp population int = 7800000000      // Hard to read
temp bytes int = 16777216             // What magnitude is this?
temp precise float = 3.14159265359    // Where do the significant digits end?
```

With underscores:
```ez
temp population int = 7_800_000_000   // Clearly 7.8 billion
temp bytes int = 16_777_216           // Easier to parse
temp precise float = 3.141_592_653_59 // Group by significance
```

## Proposed Syntax

### Integer Literals
```ez
temp million int = 1_000_000
temp billion int = 7_800_000_000
temp arbitrary int = 12_34_56         // Any grouping allowed
temp readability int = 5_000          // Improves clarity
```

### Float Literals
```ez
temp pi float = 3.141_592_653         // After decimal point
temp money float = 1_000_000.50       // Before decimal point
temp both float = 1_234.567_890       // Both sides
```

### Hexadecimal Literals
```ez
temp color int = 0xFF_AA_BB_CC        // Group by bytes
temp address int = 0xDEAD_BEEF        // Memorable patterns
temp mask int = 0x00_FF_00_FF         // Byte grouping
```

### Binary Literals
```ez
temp flags int = 0b1111_0000_1010_1100    // Nibble grouping
temp byte int = 0b1111_0000                // Clear byte boundaries
temp mask int = 0b1000_0000_0000_0001     // Bit patterns
```

### Octal Literals (if supported)
```ez
temp perms int = 0o755_644            // Unix permission grouping
```

## Placement Rules

### ✅ Allowed Placements

Underscores can appear **anywhere between digits**:
```ez
temp valid1 int = 1_000
temp valid2 int = 1_000_000
temp valid3 int = 12_34_56           // Any grouping
temp valid4 float = 3.141_592
temp valid5 float = 1_234.567_890
temp valid6 int = 0xFF_AA_BB
temp valid7 int = 0b1111_0000
```

### ❌ Forbidden Placements

#### 1. Cannot Start or End a Number
```ez
temp bad1 int = _123        // ❌ ERROR: starts with underscore
temp bad2 int = 123_        // ❌ ERROR: ends with underscore
temp bad3 float = _3.14     // ❌ ERROR: starts with underscore
temp bad4 float = 3.14_     // ❌ ERROR: ends with underscore
```

#### 2. Cannot Have Consecutive Underscores
```ez
temp bad5 int = 1__000      // ❌ ERROR: consecutive underscores
temp bad6 int = 1___000     // ❌ ERROR: consecutive underscores
```

#### 3. Cannot Be Directly Before/After Decimal Point
```ez
temp bad7 float = 3._14     // ❌ ERROR: immediately after decimal
temp bad8 float = 3_.14     // ❌ ERROR: immediately before decimal
temp bad9 float = 3_.14_    // ❌ ERROR: multiple violations
```

#### 4. Cannot Be In or Around Base Prefixes
```ez
temp bad10 int = 0_xFF      // ❌ ERROR: in hex prefix
temp bad11 int = 0x_FF      // ❌ ERROR: after hex prefix
temp bad12 int = 0_b1010    // ❌ ERROR: in binary prefix
temp bad13 int = 0b_1010    // ❌ ERROR: after binary prefix
```

## Error Messages

### Invalid Placement Errors

```ez
temp bad int = 123_

error[E1006]: invalid numeric literal
  --> test.ez:1:16
   |
1  | temp bad int = 123_
   |                ^^^^ numeric literal cannot end with underscore
   |
   = help: remove the trailing underscore or add more digits
```

```ez
temp bad float = 3._14

error[E1006]: invalid numeric literal
  --> test.ez:1:18
   |
1  | temp bad float = 3._14
   |                  ^^^^^ underscore cannot appear immediately after decimal point
   |
   = help: remove the underscore or move it between digits
```

```ez
temp bad int = 1__000

error[E1006]: invalid numeric literal
  --> test.ez:1:16
   |
1  | temp bad int = 1__000
   |                ^^^^^^ consecutive underscores not allowed in numeric literals
   |
   = help: use single underscores to separate digit groups
```

```ez
temp bad int = 0x_FF

error[E1006]: invalid numeric literal
  --> test.ez:1:16
   |
1  | temp bad int = 0x_FF
   |                ^^^^^ underscore cannot appear immediately after base prefix
   |
   = help: move the underscore to separate digit groups, not the prefix
```

## Implementation Plan

### Phase 1: Lexer Changes

Modify the number scanning logic in `pkg/lexer/lexer.go`:

1. **Update `readNumber()` function**:
   ```go
   func (l *Lexer) readNumber() string {
       position := l.position
       hasDecimal := false
       
       for {
           if isDigit(l.ch) {
               l.readChar()
           } else if l.ch == '_' {
               // Validate underscore placement
               if !isDigit(l.peekChar()) {
                   // Error: underscore not followed by digit
                   return l.input[position:l.position]
               }
               l.readChar() // consume underscore
           } else if l.ch == '.' && !hasDecimal && isDigit(l.peekChar()) {
               hasDecimal = true
               l.readChar()
           } else {
               break
           }
       }
       
       return l.input[position:l.position]
   }
   ```

2. **Add validation function**:
   ```go
   func (l *Lexer) validateNumericLiteral(literal string) error {
       // Check for leading/trailing underscores
       if strings.HasPrefix(literal, "_") {
           return fmt.Errorf("numeric literal cannot start with underscore")
       }
       if strings.HasSuffix(literal, "_") {
           return fmt.Errorf("numeric literal cannot end with underscore")
       }
       
       // Check for consecutive underscores
       if strings.Contains(literal, "__") {
           return fmt.Errorf("consecutive underscores not allowed")
       }
       
       // Check for underscores around decimal point
       if strings.Contains(literal, "_.") || strings.Contains(literal, "._") {
           return fmt.Errorf("underscore cannot be adjacent to decimal point")
       }
       
       return nil
   }
   ```

3. **Strip underscores during conversion**:
   ```go
   func (l *Lexer) tokenizeNumber(literal string) Token {
       // Validate placement
       if err := l.validateNumericLiteral(literal); err != nil {
           l.addError(errors.E1006, err.Error())
           return Token{Type: ILLEGAL, Literal: literal}
       }
       
       // Strip underscores for actual conversion
       cleaned := strings.ReplaceAll(literal, "_", "")
       
       // Determine token type (INT, FLOAT, etc.)
       // ... existing logic using 'cleaned' ...
   }
   ```

### Phase 2: Hex/Binary Literal Support

Handle underscores in different numeric bases:

```go
func (l *Lexer) readHexNumber() string {
    // Consume '0x' or '0X'
    position := l.position
    l.readChar() // '0'
    l.readChar() // 'x'
    
    for isHexDigit(l.ch) || l.ch == '_' {
        if l.ch == '_' {
            // Validate underscore placement
            if !isHexDigit(l.peekChar()) {
                break
            }
        }
        l.readChar()
    }
    
    return l.input[position:l.position]
}

func (l *Lexer) readBinaryNumber() string {
    // Similar logic for binary (0b prefix)
    // ...
}
```

### Phase 3: Error Code Addition

Add to `pkg/errors/codes.go`:

```go
// E1006 already exists for invalid numeric format
// Extend its usage to cover underscore placement errors
```

Update error messages to be specific:
- "numeric literal cannot start with underscore"
- "numeric literal cannot end with underscore"
- "consecutive underscores not allowed in numeric literals"
- "underscore cannot appear immediately after decimal point"
- "underscore cannot appear immediately before decimal point"
- "underscore cannot appear immediately after base prefix"

### Phase 4: Parser/AST Changes

**No changes needed** - the parser receives cleaned numeric strings after underscores are stripped by the lexer.

## Testing Plan

### Test Cases

#### 1. Valid Integer Separators
```ez
temp test1 int = 1_000
temp test2 int = 1_000_000
temp test3 int = 7_800_000_000
temp test4 int = 12_34_56
```

#### 2. Valid Float Separators
```ez
temp test5 float = 3.141_592_653
temp test6 float = 1_000.50
temp test7 float = 1_234.567_890
```

#### 3. Valid Hex Separators
```ez
temp test8 int = 0xFF_AA_BB_CC
temp test9 int = 0xDEAD_BEEF
temp test10 int = 0x00_11_22_33
```

#### 4. Valid Binary Separators
```ez
temp test11 int = 0b1111_0000
temp test12 int = 0b1010_1010_1010_1010
temp test13 int = 0b1111_0000_1111_0000
```

#### 5. Invalid Leading Underscore
```ez
temp bad1 int = _123           // Error
temp bad2 float = _3.14        // Error
temp bad3 int = 0x_FF          // Error (after prefix)
```

#### 6. Invalid Trailing Underscore
```ez
temp bad4 int = 123_           // Error
temp bad5 float = 3.14_        // Error
```

#### 7. Invalid Consecutive Underscores
```ez
temp bad6 int = 1__000         // Error
temp bad7 int = 1___000        // Error
temp bad8 float = 3.14__15     // Error
```

#### 8. Invalid Decimal Adjacency
```ez
temp bad9 float = 3._14        // Error
temp bad10 float = 3_.14       // Error
temp bad11 float = 3_._14      // Error
```

#### 9. Edge Cases
```ez
temp edge1 int = 1_2_3_4_5     // Valid (unusual but allowed)
temp edge2 float = 0.1_2_3     // Valid
temp edge3 int = 0xFF_F        // Valid (3 hex digits with separator)
```

#### 10. Mathematical Operations
```ez
temp calc1 int = 1_000 + 2_000              // Valid
temp calc2 float = 3.14_15 * 2.0            // Valid
temp calc3 int = 1_000_000 / 1_000          // Valid
```

## Documentation Requirements

### Language Reference Update

Add section on numeric literals:

```markdown
## Numeric Literals

### Integer Literals
Integers can be written in decimal, hexadecimal, or binary format:
- Decimal: `42`, `1000`
- Hexadecimal: `0xFF`, `0xDEADBEEF`
- Binary: `0b1010`, `0b11110000`

### Numeric Separators
Use underscores to improve readability of large numbers:
- `temp million int = 1_000_000`
- `temp pi float = 3.141_592_653`
- `temp color int = 0xFF_AA_BB_CC`

Underscores are purely cosmetic and are ignored by the compiler.

**Rules:**
- Cannot start or end a number
- Cannot be consecutive (`__`)
- Cannot be adjacent to decimal points
- Cannot be in or after base prefixes (`0x`, `0b`)
```

### Style Guide Recommendation

```markdown
## Numeric Separator Guidelines

**Recommended groupings:**
- Decimal integers: group by thousands (every 3 digits)
  - `1_000`, `1_000_000`, `1_000_000_000`
- Decimal floats: group by significance or preference
  - `3.141_592_653` (scientific constants)
  - `1_000.50` (monetary values)
- Hexadecimal: group by bytes (every 2 hex digits)
  - `0xFF_AA_BB_CC` (RGBA color)
  - `0xDEAD_BEEF` (memorable patterns)
- Binary: group by nibbles (4 bits) or bytes (8 bits)
  - `0b1111_0000` (byte)
  - `0b1111_0000_1010_1100` (16-bit value)
```

## Examples

### Before (Current)
```ez
import @std

do main() {
    using std
    
    temp population int = 7800000000
    temp pi float = 3.14159265359
    temp color int = 0xFFAABBCC
    temp flags int = 0b11110000
    
    println("Population:", population)
    println("Pi:", pi)
}
```

### After (With Separators)
```ez
import @std

do main() {
    using std
    
    temp population int = 7_800_000_000
    temp pi float = 3.141_592_653_59
    temp color int = 0xFF_AA_BB_CC
    temp flags int = 0b1111_0000
    
    println("Population:", population)
    println("Pi:", pi)
}
```

## Compatibility

This feature is **backward compatible** - existing code without underscores continues to work unchanged.

## Future Enhancements

- Linter rule to suggest underscores for large numbers (e.g., numbers > 9999)
- Formatter option to automatically add/standardize separator groupings
- IDE support for displaying numbers with separators even if source doesn't have them

## Acceptance Criteria

- [ ] Underscores work in integer literals (decimal, hex, binary, octal)
- [ ] Underscores work in float literals (both sides of decimal point)
- [ ] Leading underscores are rejected with error E1006
- [ ] Trailing underscores are rejected with error E1006
- [ ] Consecutive underscores are rejected with error E1006
- [ ] Underscores adjacent to decimal point are rejected with error E1006
- [ ] Underscores in/after base prefixes are rejected with error E1006
- [ ] Error messages are clear and helpful
- [ ] All test cases pass
- [ ] Documentation updated (language reference, style guide)
- [ ] Example programs demonstrating feature

--- a/README.md
+++ b/README.md
@@ -113,6 +113,7 @@ do add(x, y int) -> int {
 
 - **Data Types**
   - Primitives: `int`, `float`, `string`, `char`, `bool`
+  - Numeric separators: underscores for readability (`1_000_000`)
   - Arrays: dynamic `[type]` and fixed-size `[type, size]`
   - Structs: user-defined types with fields
   - Enums: integer, float, and string enums with attributes
@@ -219,6 +220,12 @@ temp price float = 19.99
 temp text string = "hello"
 temp letter char = 'A'
 temp isActive bool = true
+
+// Numeric separators for readability
+temp million int = 1_000_000
+temp billion int = 7_800_000_000
+temp pi float = 3.141_592_653
+temp money float = 1_234.56
 
 // Arrays
 temp numbers [int] = {1, 2, 3, 4, 5}

--- a/examples/numeric_separators.ez
+++ b/examples/numeric_separators.ez
@@ -1,0 +1,80 @@
+import @std
+
+do main() {
+    using std
+
+    println("=== Valid Integer Separators ===")
+
+    // Thousands separators
+    temp thousand int = 1_000
+    temp million int = 1_000_000
+    temp billion int = 7_800_000_000
+
+    println("Thousand:", thousand)
+    println("Million:", million)
+    println("Billion:", billion)
+
+    // Arbitrary grouping (all valid)
+    temp group1 int = 12_34_56
+    temp group2 int = 1_2_3_4_5
+    temp group3 int = 5_000
+
+    println("Arbitrary grouping 1:", group1)
+    println("Arbitrary grouping 2:", group2)
+    println("Arbitrary grouping 3:", group3)
+
+    println("")
+    println("=== Valid Float Separators ===")
+
+    // Float with separators
+    temp pi float = 3.141_592_653
+    temp money float = 1_000.50
+    temp both_sides float = 1_234.567_890
+    temp precise float = 3.141_592_653_59
+
+    println("Pi:", pi)
+    println("Money:", money)
+    println("Both sides:", both_sides)
+    println("Precise:", precise)
+
+    println("")
+    println("=== Arithmetic with Separators ===")
+
+    temp x int = 1_000
+    temp y int = 2_000
+    temp sum int = x + y
+    temp product int = x * y
+
+    println("1_000 + 2_000 =", sum)
+    println("1_000 * 2_000 =", product)
+
+    temp a float = 10_000.5
+    temp b float = 2_000.25
+    temp diff float = a - b
+
+    println("10_000.5 - 2_000.25 =", diff)
+
+    println("")
+    println("=== Comparisons ===")
+
+    temp population int = 7_800_000_000
+    if population > 7_000_000_000 {
+        println("Population is over 7 billion")
+    }
+
+    temp price float = 1_234.56
+    if price >= 1_000.00 {
+        println("Price is over $1000")
+    }
+
+    println("")
+    println("=== Mixed Usage ===")
+
+    temp array_size int = 10_000
+    temp bytes int = 16_777_216
+    temp percentage float = 99.999_999
+
+    println("Array size:", array_size)
+    println("Bytes:", bytes)
+    println("Percentage:", percentage)
+}

--- a/examples/numeric_separators_invalid.ez
+++ b/examples/numeric_separators_invalid.ez
@@ -1,0 +1,30 @@
+import @std
+
+do main() {
+    using std
+
+    // These should all produce errors:
+
+    // Leading underscore
+    // temp bad1 int = _123
+
+    // Trailing underscore
+    // temp bad2 int = 123_
+
+    // Consecutive underscores
+    // temp bad3 int = 1__000
+
+    // Underscore before decimal point
+    // temp bad4 float = 3_.14
+
+    // Underscore after decimal point
+    // temp bad5 float = 3._14
+
+    // Trailing underscore in float
+    // temp bad6 float = 3.14_
+
+    // Multiple violations
+    // temp bad7 float = _3._14_
+
+    println("If you see this, the error checking might not be working correctly!")
+}

--- a/pkg/lexer/lexer.go
+++ b/pkg/lexer/lexer.go
@@ -327,14 +327,62 @@ func (l *Lexer) readNumber() (string, tokenizer.TokenType) {
 	startColumn := l.column
 	tokenType := tokenizer.INT
 
-	for isDigit(l.ch) {
+	// Read integer part (digits and underscores)
+	for isDigit(l.ch) || l.ch == '_' {
+		if l.ch == '_' {
+			// Check for invalid underscore placement
+			next := l.peekChar()
+
+			// Cannot have consecutive underscores
+			if next == '_' {
+				l.addError("E1006", "consecutive underscores not allowed in numeric literals", startLine, startColumn)
+				l.readChar()
+				continue
+			}
+
+			// Cannot have underscore at end (followed by non-digit/non-underscore)
+			if !isDigit(next) && next != '.' {
+				// This might be the trailing underscore
+				l.addError("E1006", "numeric literal cannot end with underscore", startLine, startColumn)
+			}
+		}
 		l.readChar()
 	}
 
+	// Check for decimal point
 	if l.ch == '.' && isDigit(l.peekChar()) {
 		tokenType = tokenizer.FLOAT
+
+		// Check if there's an underscore before the decimal point
+		if position < l.position && l.input[l.position-1] == '_' {
+			l.addError("E1006", "underscore cannot appear immediately before decimal point", startLine, startColumn)
+		}
+
 		l.readChar() // consume '.'
-		for isDigit(l.ch) {
+
+		// Check if there's an underscore after the decimal point
+		if l.ch == '_' {
+			l.addError("E1006", "underscore cannot appear immediately after decimal point", startLine, startColumn)
+			l.readChar()
+		}
+
+		// Read fractional part (digits and underscores)
+		for isDigit(l.ch) || l.ch == '_' {
+			if l.ch == '_' {
+				next := l.peekChar()
+
+				// Cannot have consecutive underscores
+				if next == '_' {
+					l.addError("E1006", "consecutive underscores not allowed in numeric literals", startLine, startColumn)
+					l.readChar()
+					continue
+				}
+
+				// Cannot end with underscore
+				if !isDigit(next) {
+					l.addError("E1006", "numeric literal cannot end with underscore", startLine, startColumn)
+				}
+			}
 			l.readChar()
 		}
 
@@ -343,11 +391,26 @@ func (l *Lexer) readNumber() (string, tokenizer.TokenType) {
 			l.addError("E1006", "invalid number format: multiple decimal points", startLine, startColumn)
 		}
 	} else if l.ch == '.' && !isDigit(l.peekChar()) {
-		// Number ends with a dot but no digits after (e.g., "1.")
-		l.addError("E1006", "invalid number format: decimal point must be followed by digits", startLine, startColumn)
+		// Number ends with a dot but no digits after (e.g., "1." or "1._")
+		// But check if underscore before dot
+		if position < l.position && l.input[l.position-1] == '_' {
+			l.addError("E1006", "underscore cannot appear immediately before decimal point", startLine, startColumn)
+		}
+		// Check if underscore after dot
+		if l.peekChar() == '_' {
+			l.addError("E1006", "underscore cannot appear immediately after decimal point", startLine, startColumn)
+		} else {
+			l.addError("E1006", "invalid number format: decimal point must be followed by digits", startLine, startColumn)
+		}
 	}
 
-	return l.input[position:l.position], tokenType
+	// Check for leading underscore (number starts with underscore)
+	literal := l.input[position:l.position]
+	if len(literal) > 0 && literal[0] == '_' {
+		l.addError("E1006", "numeric literal cannot start with underscore", startLine, startColumn)
+	}
+
+	return literal, tokenType
 }
 
 func (l *Lexer) readString() (string, bool) {

--- a/pkg/parser/parser.go
+++ b/pkg/parser/parser.go
@@ -5,6 +5,7 @@ package parser
 import (
 	"fmt"
 	"strconv"
+	"strings"
 
 	. "github.com/marshallburns/ez/pkg/ast"
 	"github.com/marshallburns/ez/pkg/errors"
@@ -1553,7 +1554,10 @@ func (p *Parser) parseStructValue(name *Label) Expression {
 func (p *Parser) parseIntegerValue() Expression {
 	lit := &IntegerValue{Token: p.currentToken}
 
-	value, err := strconv.ParseInt(p.currentToken.Literal, 0, 64)
+	// Strip underscores for numeric conversion (they're only for readability)
+	cleanedLiteral := stripUnderscores(p.currentToken.Literal)
+
+	value, err := strconv.ParseInt(cleanedLiteral, 0, 64)
 	if err != nil {
 		msg := fmt.Sprintf("could not parse %q as integer", p.currentToken.Literal)
 		p.errors = append(p.errors, msg)
@@ -1565,10 +1569,27 @@ func (p *Parser) parseIntegerValue() Expression {
 	return lit
 }
 
+// stripUnderscores removes all underscores from a numeric literal
+func stripUnderscores(s string) string {
+	if !strings.Contains(s, "_") {
+		return s
+	}
+	var result strings.Builder
+	for _, ch := range s {
+		if ch != '_' {
+			result.WriteRune(ch)
+		}
+	}
+	return result.String()
+}
+
 func (p *Parser) parseFloatValue() Expression {
 	lit := &FloatValue{Token: p.currentToken}
 
-	value, err := strconv.ParseFloat(p.currentToken.Literal, 64)
+	// Strip underscores for numeric conversion (they're only for readability)
+	cleanedLiteral := stripUnderscores(p.currentToken.Literal)
+
+	value, err := strconv.ParseFloat(cleanedLiteral, 64)
 	if err != nil {
 		msg := fmt.Sprintf("could not parse %q as float", p.currentToken.Literal)
 		p.errors = append(p.errors, msg)

